File: hooks/scripts/internal/hookio/hookio.go
// Hook I/O helpers: stdin reading, stdout writing, error blocking.
// All errors use fail-open (exit 0) except explicit blocks (exit 2).
package hookio

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
)

// ReadInput reads JSON from stdin into target (any pointer type).
func ReadInput(target interface{}) error {
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		return err
	}
	if len(data) == 0 {
		return fmt.Errorf("empty stdin")
	}
	return json.Unmarshal(data, target)
}

// WriteOutput encodes output as JSON and writes to stdout.
func WriteOutput(output interface{}) {
	_ = json.NewEncoder(os.Stdout).Encode(output)
}

// WriteContext writes plain text to stdout (for context injection hooks).
func WriteContext(text string) {
	_, _ = os.Stdout.WriteString(text)
}

// Block writes message to stderr and exits with code 2 (explicit deny).
func Block(message string) {
	_, _ = os.Stderr.WriteString(message)
	os.Exit(2)
}

// Log writes a prefixed message to stderr (non-blocking, visible in hook output).
func Log(hookName, message string) {
	fmt.Fprintf(os.Stderr, "[%s] %s\n", hookName, message)
}

// RunHook wraps a hook handler with fail-open error handling.
// Reads JSON from stdin into target, then calls handler. Any error → exit 0.
func RunHook(hookName string, target interface{}, handler func() error) {
	if err := ReadInput(target); err != nil {
		Log(hookName, err.Error())
		os.Exit(0)
	}
	if err := handler(); err != nil {
		Log(hookName, err.Error())
		os.Exit(0)
	}
}

File: hooks/scripts/_archived_ts/lib/hook-io.ts
// Hook I/O helpers: stdin reading, stdout writing, error blocking
import * as fs from "node:fs";
import type { HookInput, HookOutput } from "./types.ts";

/** Read JSON from stdin asynchronously (works in both Node and Bun) */
export async function readInput<T extends HookInput>(): Promise<T> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer);
  }
  return JSON.parse(Buffer.concat(chunks).toString()) as T;
}

/** Read stdin synchronously (for hooks that use sync I/O) */
export function readInputSync<T extends HookInput>(): T {
  const stdin = fs.readFileSync(0, "utf-8").trim();
  if (!stdin) throw new Error("Empty stdin");
  return JSON.parse(stdin) as T;
}

/** Write JSON output to stdout */
export function writeOutput(output: HookOutput): void {
  process.stdout.write(JSON.stringify(output));
}

/** Write plain text to stdout (for SessionStart/UserPromptSubmit context) */
export function writeContext(text: string): void {
  process.stdout.write(text);
}

/** Exit with blocking error (stderr message shown to Claude, exit 2) */
export function block(message: string): never {
  process.stderr.write(message);
  process.exit(2);
}

/** Log to stderr (visible in hook output but non-blocking) */
export function log(hookName: string, message: string): void {
  process.stderr.write(`[${hookName}] ${message}\n`);
}

/** Wrap async hook in try-catch with fail-open behavior */
export async function runHook<T extends HookInput>(
  hookName: string,
  handler: (input: T) => Promise<void>
): Promise<void> {
  try {
    const input = await readInput<T>();
    await handler(input);
  } catch (err) {
    process.stderr.write(
      `[${hookName}] ${err instanceof Error ? err.message : String(err)}\n`
    );
    process.exit(0); // fail-open
  }
}

/** Wrap sync hook in try-catch with fail-open behavior */
export function runHookSync<T extends HookInput>(
  hookName: string,
  handler: (input: T) => void
): void {
  try {
    const input = readInputSync<T>();
    handler(input);
  } catch (err) {
    process.stderr.write(
      `[${hookName}] ${err instanceof Error ? err.message : String(err)}\n`
    );
    process.exit(0);
  }
}
